pkg/server/plugin/upstreamca/awssecret/awssecret.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/spiffe/spire/proto/server/upstreamca"
+)
+
+const (
+	pluginName = "aws_secrets"
+)
+
+type AWSSecretConfiguration struct {
+	TTL                   string `hcl:"ttl" json:"ttl"` // time to live for generated certs
+	Region                string `hcl:"region" json:"region"`
+	CertFileARN           string `hcl:"cert_file_arn" json:"cert_file_arn"`
+	KeyFileARN            string `hcl:"key_file_arn" json:"key_file_arn"`
+	AccessKeyID           string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey       string `hcl:"secret_access_key" json:"secret_access_key"`
+	SecurityToken         string `hcl:"secret_token" json:"secret_token"`
+	UseFakeSecretsManager string `hcl:"use_fake_secretsmanager" json:"use_fake_secretsmanager"`
+}
+
+type awssecretPlugin struct {
+	serialNumber x509util.SerialNumber
+
+	mtx        sync.RWMutex
+	cert       *x509.Certificate
+	upstreamCA *x509svid.UpstreamCA
+
+	mu                sync.RWMutex
+	config            *AWSSecretConfiguration
+	secrm             secretsmanageriface.SecretsManagerAPI
+	setSecretsManager func(secretsmanageriface.SecretsManagerAPI)
+
+	hooks struct {
+		getenv            func(string) string
+		newClient         func(config *AWSSecretConfiguration, region string) (secretsmanageriface.SecretsManagerAPI, error)
+		setSecretsManager func(s secretsmanageriface.SecretsManagerAPI)
+	}
+}
+
+func fetchFromSecretsManager(config *AWSSecretConfiguration, sm secretsmanageriface.SecretsManagerAPI) (*ecdsa.PrivateKey, *x509.Certificate, error) {
+
+	keyPEMstr, err := readARN(sm, config.KeyFileARN)
+
+	if keyPEMstr == nil || err != nil {
+		return nil, nil, fmt.Errorf("unable to read %s: %s", config.KeyFileARN, err)
+	}
+
+	keyPEM := []byte(*keyPEMstr)
+
+	block, rest := pem.Decode(keyPEM)
+
+	if block == nil {
+		return nil, nil, errors.New("invalid key format")
+	}
+
+	if len(rest) > 0 {
+		return nil, nil, errors.New("invalid key format: too many keys")
+	}
+
+	key, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEMstr, err := readARN(sm, config.CertFileARN)
+
+	if certPEMstr == nil || err != nil {
+		return nil, nil, fmt.Errorf("unable to read %s: %s", config.CertFileARN, err)
+	}
+
+	certPEM := []byte(*certPEMstr)
+
+	block, rest = pem.Decode(certPEM)
+
+	if block == nil {
+		return nil, nil, errors.New("invalid cert format")
+	}
+
+	if len(rest) > 0 {
+		return nil, nil, errors.New("invalid cert format: too many certs")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Validate cert matches private key
+	matched, err := x509util.CertificateMatchesPrivateKey(cert, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !matched {
+		return nil, nil, errors.New("certificate and private key does not match")
+	}
+
+	return key, cert, nil
+}
+
+func (m *awssecretPlugin) configValidation(ctx context.Context, req *spi.ConfigureRequest) (*AWSSecretConfiguration, error) {
```

Comment:
nit: seems like `validateConfig` or `parseConfig` would be a more appropriate name.
also nit: unexported struct methods belong below the exported struct methods.